sidecar/src/watchdog.rs

```diff
@@ -0,0 +1,198 @@
+                    _ = self.interval.tick() => {
+                        let current_mem_usage_bytes = memory_stats::memory_stats()
+                        .map(|s| s.physical_mem)
+                        .unwrap_or(0);
+                        mem_usage_bytes.store(current_mem_usage_bytes, Ordering::Relaxed);
+
+                        if current_mem_usage_bytes > self.max_memory_usage_bytes {
+                            std::thread::spawn(||{
+                                // TODO: we should trigger manual flush and submission here
+                                // wait 5 seconds to give metrics a chance to flush - then kill the process
+                                std::thread::sleep(Duration::from_secs(5)); 
+                                std::process::exit(1);
+                            });
+                            return
+                        }
+                        
+                    },
+                    _ = self.shutdown_receiver.recv() => {
+                        return
+                    },
+                }
+            }
+        });
+        WatchdogHandle {
+            handle: join_handle.map(Result::ok).boxed().shared(),
+            mem_usage_bytes: handle_mem_usage_bytes,
+        }
+    }
+}
+
+pub struct SelfTelemetry {
+    pub submission_interval: tokio::time::Interval,
+    pub watchdog_handle: WatchdogHandle,
+    pub config: ddtelemetry::config::Config,
+    pub server: SidecarServer,
+}
+
+impl SelfTelemetry {
+    /// spawn_worker
+    ///
+    /// should always succeed
+    /// not to bring down other functionality if we fail to initialize the internal telemetry
+    pub async fn spawn_worker(mut self) {
+        let (worker, join_handle) = match TelemetryWorkerBuilder::new_fetch_host(
+            "datadog-ipc-helper".to_string(),
+            "php".to_string(),
+            "SIDECAR".to_string(),
+            env!("CARGO_PKG_VERSION").to_string(),
+        )
+        .spawn_with_config(self.config.clone())
+        .await
+        {
+            Ok(r) => r,
+            Err(_err) => {
+                self.watchdog_handle.wait_for_shutdown().await;
+                return;
+            }
+        };
+
+        let metrics = MetricData {
+            worker: &worker,
+            server: &self.server,
+            sidecar_watchdog: &self.watchdog_handle,
+            submitted_payloads: worker.register_metric_context(
+                "server.submitted_payloads".to_string(),
+                vec![],
+                MetricType::Count,
+                true,
+                MetricNamespace::Sidecar,
+            ),
+            active_sessions: worker.register_metric_context(
+                "server.active_sessions".to_string(),
+                vec![],
+                MetricType::Gauge,
+                true,
+                MetricNamespace::Sidecar,
+            ),
+            memory_usage: worker.register_metric_context(
+                "server.memory_usage".to_string(),
```

Comment:
@paullegranddc Hey Paul, do we have to add this new metric somewhere ?

Reply:
Yes, you would have to add it here https://github.com/DataDog/dd-go/blob/b52550e0ca6462443f66286bc717dc635947a365/trace/apps/tracer-telemetry-intake/telemetry-metrics/static/common_metrics.json